prov/hook/trace/src/hook_trace.c

```diff
@@ -1064,11 +1085,329 @@ struct fi_ops_cq trace_cq_ops = {
 	.strerror = hook_cq_strerror,
 };
 
-static int trace_cq_init(struct fid *fid)
+/* av ops */
+static int
+trace_av_insert(struct fid_av *av, const void *addr, size_t count,
+                fi_addr_t *fi_addr, uint64_t flags, void *context)
+{
+	struct hook_av *myav = container_of(av, struct hook_av, av);
+
+	return fi_av_insert(myav->hav, addr, count, fi_addr, flags, context);
+}
+
+static int
+trace_av_insertsvc(struct fid_av *av, const char *node,
+                   const char *service, fi_addr_t *fi_addr, uint64_t flags,
+                   void *context)
+{
+	struct hook_av *myav = container_of(av, struct hook_av, av);
+
+	return fi_av_insertsvc(myav->hav, node, service, fi_addr, flags,
+                           context);
+}
+
+static int
+trace_av_insertsym(struct fid_av *av, const char *node, size_t nodecnt,
+                   const char *service, size_t svccnt, fi_addr_t *fi_addr,
+                   uint64_t flags, void *context)
+{
+	struct hook_av *myav = container_of(av, struct hook_av, av);
+
+	return fi_av_insertsym(myav->hav, node, nodecnt, service, svccnt,
+	                       fi_addr, flags, context);
+}
+
+static int
+trace_av_remove(struct fid_av *av, fi_addr_t *fi_addr, size_t count,
+                uint64_t flags)
+{
+	struct hook_av *myav = container_of(av, struct hook_av, av);
+
+	return fi_av_remove(myav->hav, fi_addr, count, flags);
+}
+
+static int
+trace_av_lookup(struct fid_av *av, fi_addr_t fi_addr, void *addr,
+                size_t *addrlen)
+{
+	struct hook_av *myav = container_of(av, struct hook_av, av);
+
+	return fi_av_lookup(myav->hav, fi_addr, addr, addrlen);
+}
+
+static const char *
+trace_av_straddr(struct fid_av *av, const void *addr, char *buf, size_t *len)
+{
+	struct hook_av *myav = container_of(av, struct hook_av, av);
+
+	return fi_av_straddr(myav->hav, addr, buf, len);
+}
+
+static struct fi_ops_av trace_av_ops = {
+	.size = sizeof(struct fi_ops_av),
+	.insert = trace_av_insert,
+	.insertsvc = trace_av_insertsvc,
+	.insertsym = trace_av_insertsym,
+	.remove = trace_av_remove,
+	.lookup = trace_av_lookup,
+	.straddr = trace_av_straddr,
+};
+
+/* domain ops */
+static int
+trace_av_open(struct fid_domain *domain, struct fi_av_attr *attr,
+              struct fid_av **av, void *context)
+{
+	struct hook_domain *dom = container_of(domain, struct hook_domain, domain);
+	struct hook_av *myav;
+	char buf[TRACE_BUF_SIZE];
+	int ret;
+
+	myav = calloc(1, sizeof *myav);
+	if (!myav)
+		return -FI_ENOMEM;
+
+	myav->domain = dom;
+	myav->av.fid.fclass = FI_CLASS_AV;
+	myav->av.fid.context = context;
+	myav->av.fid.ops = &hook_fid_ops;
+	myav->av.ops = &trace_av_ops;
+
+	ret = fi_av_open(dom->hdomain, attr, &myav->hav, &myav->av.fid);
+	if (!ret) {
+		FI_TRACE(dom->fabric->hprov, FI_LOG_DOMAIN,
+		        "av %p, context %p\n"
+		        "%s",
```

Comment:
nit: I'd toss the %s into the previous line to avoid having a line dedicated to just that format string and tighten the code.  Same comment applies to other functions below.